File: apps/api/src/modules/scans/scans.service.ts
import { Injectable } from '@nestjs/common';
import { Prisma, type Product } from '@prisma/client';
import type {
  BarcodeLookupProduct,
  ScanDetailResponse,
  ScanHistoryItem,
  ScanHistoryResponse,
  SharedScanFilters,
} from '@acme/shared';
import {
  normalizedProductSchema,
  productAnalysisResultSchema,
  productComparisonResultSchema,
} from '@acme/shared';
import { ApiError } from '../../shared/errors/api-error';
import {
  getProductImageUrl,
  resolveCanonicalProductImageUrl,
} from '../../shared/utils/product-image';
import { buildProductSearchFilter, normalizeSearchQuery } from '../../shared/utils/product-search';
import { prisma } from '../product-analyze/lib/prisma';
import { productFactsAiOutputSchema } from '../product-analyze/domain/product-facts/schema';
import { toBarcodeLookupProduct } from '../product-analyze/utils/analysis-response.utils';
import { buildHistoryAnalysisSummary, matchesSharedScanFilters } from './scan-history-analysis';

const DEFAULT_PAGE_SIZE = 20;
const FILTERED_BATCH_MULTIPLIER = 3;

type HistoryProduct = {
  id: string;
  barcode: string;
  product_name: string | null;
  brands: string | null;
  image_url: string | null;
  images: unknown;
  nutriscore_grade: string | null;
  classificationCache: Prisma.JsonValue | null;
};

type ProductScanHistoryRecord = {
  id: string;
  type: 'product' | 'comparison';
  personalAnalysisJobId: string | null;
  createdAt: Date;
  source: 'barcode' | 'photo';
  overallScore: number | null;
  overallRating: string | null;
  personalAnalysisStatus: 'pending' | 'completed' | 'failed' | null;
  personalResult: unknown;
  multiProfileResult: unknown;
  product: HistoryProduct | null;
  product2: HistoryProduct | null;
};

const serializeHistoryProduct = (product: HistoryProduct) => {
  const classification = productFactsAiOutputSchema.safeParse(product.classificationCache).data;

  return {
    id: product.id,
    barcode: product.barcode,
    product_name: product.product_name,
    brands: product.brands,
    image_url: resolveCanonicalProductImageUrl(product.image_url, product.images),
    nutriscore_grade: product.nutriscore_grade,
    dietCompatibility: classification?.dietCompatibility,
  };
};

const serializeDetailProduct = (
  product: Product,
  analysisResult?: Prisma.JsonValue | null,
): BarcodeLookupProduct | null => {
  const canonicalImageUrl = resolveCanonicalProductImageUrl(product.image_url, product.images);
  const classification = productFactsAiOutputSchema.safeParse(product.classificationCache).data;
  const parsedAnalysisResult = analysisResult
    ? productAnalysisResultSchema.safeParse(analysisResult).data
    : undefined;

  const parsedProduct = normalizedProductSchema.safeParse({
    code: product.code,
    product_name: product.product_name,
    brands: product.brands,
    image_url: canonicalImageUrl,
    ingredients_text: product.ingredients_text,
    nutriscore_grade: product.nutriscore_grade,
    categories: product.categories,
    quantity: product.quantity,
    serving_size: product.serving_size,
    ingredients: product.ingredients,
    allergens: product.allergens,
    additives: product.additives,
    additives_count: product.additives_count,
    traces: product.traces,
    countries: product.countries,
    category_tags: product.category_tags,
    images: {
      front_url: canonicalImageUrl,
      ingredients_url: getProductImageUrl(product.images, 'ingredients_url'),
      nutrition_url: getProductImageUrl(product.images, 'nutrition_url'),
    },
    nutrition: product.nutrition,
    scores: product.scores,
  });

  return parsedProduct.success
    ? toBarcodeLookupProduct(
        parsedProduct.data,
        classification?.dietCompatibility ?? parsedAnalysisResult?.productFacts.dietCompatibility,
      )
    : null;
};

const getValidLimit = (limit?: string): number | undefined => {
  if (!limit) {
    return undefined;
  }

  const parsed = Number.parseInt(limit, 10);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : undefined;
};

const hasSharedScanFilters = (filters: SharedScanFilters): boolean =>
  filters.profileIds.length > 0 || filters.fitBuckets.length > 0;

@Injectable()
export class ScansService {
  async getHistory(
    userId: string,
    cursor?: string,
    limit?: string,
    search?: string,
    filters: SharedScanFilters = { profileIds: [], fitBuckets: [] },
  ): Promise<ScanHistoryResponse> {
    const take = getValidLimit(limit) ?? DEFAULT_PAGE_SIZE;
    const normalizedSearch = normalizeSearchQuery(search);
    const where: Prisma.ScanWhereInput = {
      userId,
      type: 'product',
      ...(normalizedSearch
        ? {
            product: {
              is: buildProductSearchFilter(normalizedSearch),
            },
          }
        : {}),
    };

    const [scans, totalCount] = await Promise.all([
      hasSharedScanFilters(filters)
        ? this.findFilteredHistoryScans(where, cursor, take + 1, filters)
        : this.findHistoryScans(where, cursor, take + 1),
      this.countHistory(where, filters),
    ]);

    const hasMore = scans.length > take;
    const items = hasMore ? scans.slice(0, take) : scans;
    const nextCursor = hasMore ? (items[items.length - 1]?.id ?? null) : null;

    const productIds = items
      .map((scan) => scan.product?.id)
      .filter((productId): productId is string => productId != null);
    const favouriteSet = await this.getFavouriteProductIds(userId, productIds);

    return {
      items: items.map((scan) => this.serializeHistoryItem(scan, favouriteSet)),
      nextCursor,
      totalCount,
    };
  }

  private async countHistory(
    where: Prisma.ScanWhereInput,
    filters: SharedScanFilters,
  ): Promise<number> {
    if (!hasSharedScanFilters(filters)) {
      return prisma.scan.count({ where });
    }

    let totalCount = 0;
    let nextCursor: string | undefined;
    const batchSize = 100;

    while (true) {
      const scans = await prisma.scan.findMany({
        where,
        orderBy: { createdAt: 'desc' },
        take: batchSize,
        ...(nextCursor
          ? {
              cursor: { id: nextCursor },
              skip: 1,
            }
          : {}),
        select: {
          id: true,
          personalResult: true,
          multiProfileResult: true,
        },
      });

      if (scans.length === 0) {
        break;
      }

      totalCount += scans.filter((scan) => {
        const summary = buildHistoryAnalysisSummary(scan.personalResult, scan.multiProfileResult);
        return matchesSharedScanFilters(summary, filters);
      }).length;

      if (scans.length < batchSize) {
        break;
      }

      nextCursor = scans[scans.length - 1]?.id;
    }

    return totalCount;
  }

  private async findHistoryScans(
    where: Prisma.ScanWhereInput,
    cursor: string | undefined,
    take: number,
  ): Promise<ProductScanHistoryRecord[]> {
    return prisma.scan.findMany({
      where,
      orderBy: { createdAt: 'desc' },
      take,
      ...(cursor
        ? {
            cursor: { id: cursor },
            skip: 1,
          }
        : {}),
      include: {
        product: {
          select: {
            id: true,
            barcode: true,
            product_name: true,
            brands: true,
            image_url: true,
            images: true,
            nutriscore_grade: true,
            classificationCache: true,
          },
        },
        product2: {
          select: {
            id: true,
            barcode: true,
            product_name: true,
            brands: true,
            image_url: true,
            images: true,
            nutriscore_grade: true,
            classificationCache: true,
          },
        },
      },
    });
  }

  private async findFilteredHistoryScans(
    where: Prisma.ScanWhereInput,
    cursor: string | undefined,
    take: number,
    filters: SharedScanFilters,
  ): Promise<ProductScanHistoryRecord[]> {
    const matches: ProductScanHistoryRecord[] = [];
    const batchSize = Math.max(take * FILTERED_BATCH_MULTIPLIER, DEFAULT_PAGE_SIZE);
    let nextCursor = cursor;

    while (matches.length < take) {
      const scans = await this.findHistoryScans(where, nextCursor, batchSize);

      if (scans.length === 0) {
        break;
      }

      for (const scan of scans) {
        const summary = buildHistoryAnalysisSummary(scan.personalResult, scan.multiProfileResult);

        if (matchesSharedScanFilters(summary, filters)) {
          matches.push(scan);
        }

        if (matches.length >= take) {
          break;
        }
      }

      if (scans.length < batchSize) {
        break;
      }

      nextCursor = scans[scans.length - 1]?.id;
    }

    return matches;
  }

  async getDetail(userId: string, scanId: string): Promise<ScanDetailResponse> {
    const scan = await prisma.scan.findFirst({
      where: { id: scanId, userId },
      include: {
        product: true,
        product2: true,
      },
    });

    if (!scan) {
      throw ApiError.notFound('Scan not found');
    }

    const product = scan.product ? serializeDetailProduct(scan.product, scan.personalResult) : null;

    const analysisResult = scan.personalResult
      ? (productAnalysisResultSchema.safeParse(scan.personalResult).data ?? null)
      : null;

    const comparisonResult =
      scan.type === 'comparison' && scan.comparisonResult
        ? (productComparisonResultSchema.safeParse(scan.comparisonResult).data ?? null)
        : null;

    const isFavourite = scan.productId ? await this.isFavourite(userId, scan.productId) : false;

    return {
      id: scan.id,
      type: scan.type,
      analysisId: scan.personalAnalysisJobId ?? null,
      createdAt: scan.createdAt.toISOString(),
      source: scan.source,
      overallScore: scan.overallScore,
      overallRating: scan.overallRating,
      personalAnalysisStatus: scan.personalAnalysisStatus,
      barcode: scan.barcode,
      productId: scan.productId ?? null,
      isFavourite,
      product,
      analysisResult,
      comparisonResult,
    };
  }

  async deleteScan(userId: string, scanId: string): Promise<void> {
    const scan = await prisma.scan.findFirst({
      where: { id: scanId, userId },
      select: {
        id: true,
        type: true,
        productId: true,
      },
    });

    if (!scan) {
      throw ApiError.notFound('History entry not found');
    }

    await prisma.scan.delete({ where: { id: scan.id } });

    if (scan.type !== 'product' || !scan.productId) {
      return;
    }

    const remainingProductScanCount = await prisma.scan.count({
      where: {
        userId,
        type: 'product',
        productId: scan.productId,
      },
    });

    if (remainingProductScanCount > 0) {
      return;
    }

    await prisma.favorite.deleteMany({
      where: {
        userId,
        productId: scan.productId,
      },
    });

    console.log(
      `[scans] Removed favourite for productId=${scan.productId} after deleting the last product scan`,
    );
  }

  private serializeHistoryItem(
    scan: ProductScanHistoryRecord,
    favouriteSet: Set<string>,
  ): ScanHistoryItem {
    const analysisSummary = buildHistoryAnalysisSummary(
      scan.personalResult,
      scan.multiProfileResult,
    );

    return {
      id: scan.id,
      type: scan.type,
      analysisId: scan.personalAnalysisJobId ?? null,
      createdAt: scan.createdAt.toISOString(),
      source: scan.source,
      overallScore: scan.overallScore,
      overallRating: scan.overallRating,
      personalScore: analysisSummary.personalScore,
      personalRating: analysisSummary.personalRating,
      personalAnalysisStatus: scan.personalAnalysisStatus,
      mainUserHasDietConflict: analysisSummary.mainUserHasDietConflict,
      mainUserHasAllergenConflict: analysisSummary.mainUserHasAllergenConflict,
      isFavourite: scan.product ? favouriteSet.has(scan.product.id) : false,
      profileChips: analysisSummary.profileChips,
      product: scan.product ? serializeHistoryProduct(scan.product) : null,
      product2: scan.product2 ? serializeHistoryProduct(scan.product2) : null,
    };
  }

  private async isFavourite(userId: string, productId: string): Promise<boolean> {
    const favourite = await prisma.favorite.findUnique({
      where: { userId_productId: { userId, productId } },
      select: { id: true },
    });

    return favourite != null;
  }

  private async getFavouriteProductIds(userId: string, productIds: string[]): Promise<Set<string>> {
    if (productIds.length === 0) {
      return new Set();
    }

    const favourites = await prisma.favorite.findMany({
      where: { userId, productId: { in: productIds } },
      select: { productId: true },
    });

    return new Set(favourites.map((favourite) => favourite.productId));
  }
}

File: apps/server-py/app/services/scans_service.py
from typing import Optional

from sqlalchemy import and_, or_, select
from sqlalchemy.orm import selectinload

from app.enums import ScanType
from app.models.product import Product
from app.models.scan import Scan
from app.schemas.scans import (
    ProfileChip,
    ProductSummary,
    ScanDetailResponse,
    ScanHistoryItem,
    ScanHistoryResponse,
)
from app.utils.unitofwork import UnitOfWork

DEFAULT_PAGE_SIZE = 20


def _product_summary(product: Optional[Product]) -> Optional[ProductSummary]:
    if product is None:
        return None
    return ProductSummary(
        id=product.id,
        barcode=product.barcode,
        product_name=product.product_name,
        brands=product.brands,
        image_url=product.image_url,
        nutriscore_grade=product.nutriscore_grade,
    )


def _extract_profile_chips(multi_profile_result: Optional[dict]) -> Optional[list[ProfileChip]]:
    if not multi_profile_result or not isinstance(multi_profile_result, dict):
        return None
    profiles = multi_profile_result.get("profiles")
    if not isinstance(profiles, list) or not profiles:
        return None
    chips = []
    for p in profiles:
        if not isinstance(p, dict):
            continue
        profile_id = p.get("profileId")
        name = p.get("name")
        if profile_id and name:
            chips.append(
                ProfileChip(
                    profileId=profile_id,
                    name=name,
                    score=p.get("score"),
                    fitLabel=p.get("fitLabel"),
                )
            )
    return chips or None


class ScansService:
    async def get_history(
        self,
        uow: UnitOfWork,
        *,
        user_id: str,
        cursor: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> ScanHistoryResponse:
        take = limit or DEFAULT_PAGE_SIZE

        async with uow:
            stmt = (
                select(Scan)
                .where(Scan.user_id == user_id, Scan.type == ScanType.product)
                .options(
                    selectinload(Scan.product),
                    selectinload(Scan.product2),
                )
                .order_by(Scan.created_at.desc(), Scan.id.desc())
                .limit(take + 1)
            )

            if cursor:
                cursor_stmt = select(Scan.created_at, Scan.id).where(Scan.id == cursor)
                cursor_row = (await uow.session.execute(cursor_stmt)).one_or_none()
                if cursor_row:
                    stmt = stmt.where(
                        or_(
                            Scan.created_at < cursor_row.created_at,
                            and_(
                                Scan.created_at == cursor_row.created_at,
                                Scan.id < cursor,
                            ),
                        )
                    )

            result = await uow.session.execute(stmt)
            scans = list(result.scalars().all())

            # Collect product ids and check favourites
            product_ids = [s.product.id for s in scans if s.product]
            favourite_set: set[str] = set()
            if product_ids:
                from app.models.favorite import Favorite

                fav_stmt = select(Favorite.product_id).where(
                    Favorite.user_id == user_id,
                    Favorite.product_id.in_(product_ids),
                )
                fav_result = await uow.session.execute(fav_stmt)
                favourite_set = set(fav_result.scalars().all())

        has_more = len(scans) > take
        items = scans[:take]
        next_cursor = items[-1].id if has_more and items else None

        history_items = []
        for scan in items:
            raw = scan.personal_result or {}
            profiles = raw.get("profiles", []) if isinstance(raw, dict) else []
            first = profiles[0] if profiles else {}
            personal_score = first.get("score") if isinstance(first, dict) else None
            personal_rating = first.get("fitLabel") if isinstance(first, dict) else None

            history_items.append(
                ScanHistoryItem(
                    id=scan.id,
                    type=scan.type,
                    createdAt=scan.created_at.isoformat(),
                    source=scan.source,
                    overallScore=scan.overall_score,
                    overallRating=scan.overall_rating,
                    personalScore=personal_score,
                    personalRating=personal_rating,
                    personalAnalysisStatus=scan.personal_analysis_status,
                    isFavourite=bool(scan.product and scan.product.id in favourite_set),
                    profileChips=_extract_profile_chips(scan.multi_profile_result),
                    product=_product_summary(scan.product),
                    product2=_product_summary(scan.product2),
                )
            )

        return ScanHistoryResponse(items=history_items, nextCursor=next_cursor)

    async def get_scan(
        self,
        uow: UnitOfWork,
        *,
        scan_id: str,
        user_id: str,
    ) -> Optional[ScanDetailResponse]:
        async with uow:
            stmt = (
                select(Scan)
                .where(Scan.id == scan_id, Scan.user_id == user_id)
                .options(
                    selectinload(Scan.product),
                    selectinload(Scan.product2),
                )
            )
            result = await uow.session.execute(stmt)
            scan = result.scalars().one_or_none()

            if scan is None:
                return None

            is_fav = False
            if scan.product_id:
                from app.models.favorite import Favorite

                fav_stmt = select(Favorite.id).where(
                    Favorite.user_id == user_id,
                    Favorite.product_id == scan.product_id,
                )
                fav_result = await uow.session.execute(fav_stmt)
                is_fav = fav_result.scalar_one_or_none() is not None

        # Build product data dict if available
        product_data = None
        if scan.product:
            p = scan.product
            product_data = {
                "code": p.code,
                "product_name": p.product_name,
                "brands": p.brands,
                "image_url": p.image_url,
                "ingredients_text": p.ingredients_text,
                "nutriscore_grade": p.nutriscore_grade,
                "categories": p.categories,
                "quantity": p.quantity,
                "serving_size": p.serving_size,
                "ingredients": p.ingredients,
                "allergens": p.allergens,
                "additives": p.additives,
                "additives_count": p.additives_count,
                "traces": p.traces,
                "countries": p.countries,
                "category_tags": p.category_tags,
                "images": p.images,
                "nutrition": p.nutrition,
                "scores": p.scores,
            }

        return ScanDetailResponse(
            id=scan.id,
            type=scan.type,
            analysisId=scan.personal_analysis_job_id,
            createdAt=scan.created_at.isoformat(),
            source=scan.source,
            overallScore=scan.overall_score,
            overallRating=scan.overall_rating,
            personalAnalysisStatus=scan.personal_analysis_status,
            barcode=scan.barcode,
            productId=scan.product_id,
            isFavourite=is_fav,
            product=product_data,
            analysisResult=scan.personal_result,
            comparisonResult=scan.comparison_result if scan.type == ScanType.comparison else None,
        )
